File: DawnwalkerConvai-0.30.8-Release/Source/bridge/companion-orders.mjs
/** Bounded orders only. This module never evaluates text or calls game APIs. */
export const ORDER_GRAMMAR = 'Separate steps with "then". Commands: follow me / regroup; hold position; attack my target; frontline / rearguard / flank left / flank right / prefer ranged; be passive / defensive / aggressive; protect me; wait 1–120 seconds. Conditions: when combat starts, when the target is clear, or when your health is below 1–99 percent. Conditions wait up to 60 game seconds. Characters choose their own abilities. Unknown commands are rejected.';
const MAX_NODES = 24, MAX_TRANSITIONS = 96, MAX_WIRE = 12288;
const ID = /^[A-Za-z][A-Za-z0-9_-]{0,47}$/;
const TERMINALS = new Set(['done', 'failed']);
const VALUES = {
  follow: ['player'], hold: ['position'], attack: ['current_target'],
  role: ['frontline', 'rearguard', 'flank_left', 'flank_right', 'ranged'],
  stance: ['passive', 'defensive', 'aggressive'],
  priority: ['player_target','protect','nearest','weakest'],
};
const fail = message => { throw new Error(`Companion order: ${message}`); };
const integer = (n, lo, hi) => Number.isSafeInteger(n) && n >= lo && n <= hi;
function validValue(op, value, timeoutMs) {
  if (typeof value !== 'string') return false;
  if (Object.hasOwn(VALUES, op)) return VALUES[op].includes(value);
  if (op === 'wait') return /^[1-9]\d{0,5}$/.test(value) && integer(Number(value), 1, 120000) && Number(value) <= timeoutMs;
  if (op === 'condition') return value === 'combat' || value === 'target_clear' || /^health_below:0\.\d{1,3}$/.test(value) && Number(value.split(':')[1]) > 0;
  return false;
}

/** Returns a detached, canonical graph. Every branch must have an explicit edge. */
export function validateGraph(input) {
  if (!input || input.version !== 1 || !ID.test(input.id ?? '') || TERMINALS.has(input.id)) fail('invalid graph version or id');
  if (!integer(input.maxTransitions, 1, MAX_TRANSITIONS)) fail('maxTransitions must explicitly bound execution to 1–96 steps');
  if (!Array.isArray(input.nodes) || input.nodes.length < 1 || input.nodes.length > MAX_NODES) fail('provide 1–24 nodes');
  const nodes = [], byId = new Map();
  for (const n of input.nodes) {
    if (!n || !ID.test(n.id ?? '') || TERMINALS.has(n.id) || byId.has(n.id)) fail('node ids must be unique bounded identifiers');
    if (!integer(n.timeoutMs, 1, 120000) || !validValue(n.op, n.value, n.timeoutMs)) fail(`unsupported operation/value or timeout at ${n.id}`);
    if (typeof n.success !== 'string' || typeof n.failure !== 'string') fail(`both success and failure edges are required at ${n.id}`);
    const node = {id:n.id, op:n.op, value:n.value, timeoutMs:n.timeoutMs, success:n.success, failure:n.failure};
    byId.set(node.id, node); nodes.push(node);
  }
  if (!byId.has(input.start)) fail('start must identify an existing node');
  for (const n of nodes) for (const edge of [n.success, n.failure]) if (!TERMINALS.has(edge) && !byId.has(edge)) fail(`missing edge ${edge} at ${n.id}`);
  const reachable = new Set(), visit = id => {
    if (TERMINALS.has(id) || reachable.has(id)) return;
    reachable.add(id); const n = byId.get(id); visit(n.success); visit(n.failure);
  };
  visit(input.start);
  if (reachable.size !== nodes.length) fail('unreachable nodes are not allowed');
  const exits = new Set(TERMINALS);
  for (let round = 0; round < nodes.length; round++) for (const n of nodes) if (exits.has(n.success) || exits.has(n.failure)) exits.add(n.id);
  if (nodes.some(n => !exits.has(n.id))) fail('every node needs a path to a terminal; closed loops are forbidden');
  return {version:1, id:input.id, start:input.start, maxTransitions:input.maxTransitions, nodes};
}

/** ASCII TSV; identifiers and values cannot contain tabs/newlines/escapes. */
export function encodeGraph(input) {
  const g = validateGraph(input);
  return [`ORDER\t1\t${g.id}\t${g.start}\t${g.maxTransitions}`, ...g.nodes.map(n => ['NODE', n.id, n.op, n.value, n.timeoutMs, n.success, n.failure].join('\t'))].join('\n') + '\n';
}

export function decodeGraph(text) {
  if (typeof text !== 'string' || text.length > MAX_WIRE || /[^\x09\x0a\x0d\x20-\x7e]/.test(text)) fail('invalid or oversized wire data');
  const lines = text.replace(/\r\n/g, '\n').replace(/\n$/, '').split('\n');
  const h = lines.shift().split('\t');
  if (h.length !== 5 || h[0] !== 'ORDER' || h[1] !== '1' || !/^[1-9]\d?$/.test(h[4])) fail('invalid wire header');
  const nodes = lines.map(line => {
    const p = line.split('\t');
    if (p.length !== 7 || p[0] !== 'NODE' || !/^[1-9]\d{0,5}$/.test(p[4])) fail('invalid wire node');
    return {id:p[1], op:p[2], value:p[3], timeoutMs:Number(p[4]), success:p[5], failure:p[6]};
  });
  return validateGraph({version:1, id:h[2], start:h[3], maxTransitions:Number(h[4]), nodes});
}

function condition(text) {
  if (/^(?:combat starts|in combat)$/.test(text)) return 'combat';
  if (/^(?:the )?target is clear$/.test(text)) return 'target_clear';
  const health = text.match(/^(?:your |companion )health is below ([1-9]\d?) (?:percent|%)$/);
  if (health) return `health_below:${Number(health[1]) / 100}`;
  fail(`unsupported condition "${text}". ${ORDER_GRAMMAR}`);
}
function command(text) {
  const commands = [
    [/^(?:follow(?: me)?|regroup(?: with me)?)$/, 'follow', 'player'],
    [/^(?:hold(?: position)?|stop walking|stay here)$/, 'hold', 'position'],
    [/^attack (?:my|the current) target$/, 'attack', 'current_target'],
    [/^(?:be |take the |act as )?(?:a )?(?:frontline|front line)(?: fighter| role)?$/, 'role', 'frontline'],
    [/^(?:be |take the |act as )?(?:a )?(?:rearguard|rear guard)(?: fighter| role)?$/, 'role', 'rearguard'],
    [/^flank left$/, 'role', 'flank_left'], [/^flank right$/, 'role', 'flank_right'],
    [/^(?:prefer ranged|ranged preference|keep your distance)$/, 'role', 'ranged'],
    [/^protect (?:me|the player)$/, 'priority', 'protect'],
  ];
  for (const [pattern, op, value] of commands) if (pattern.test(text)) return {op, value, timeoutMs:15000};
  const stance = text.match(/^(?:be |fight |become )?(passive|defensive|aggressive)$/);
  if (stance) return {op:'stance', value:stance[1], timeoutMs:15000};
  const wait = text.match(/^wait ([1-9]\d{0,2})(?: seconds?|s)$/);
  if (wait && Number(wait[1]) <= 120) return {op:'wait', value:String(Number(wait[1]) * 1000), timeoutMs:Math.min(120000, Number(wait[1]) * 1000 + 1000)};
  fail(`unsupported command "${text}". ${ORDER_GRAMMAR}`);
}

/** Deterministic small grammar, not a general natural-language planner. */
export function compileOrder(text, {id = 'order'} = {}) {
  if (typeof text !== 'string' || text.length < 1 || text.length > 2048 || /[\x00-\x1f\x7f]/.test(text)) fail('use a plain order of at most 2048 characters');
  const normalized = text.trim().toLowerCase().replace(/[.!]$/, '').replace(/\s+/g, ' ');
  const clauses = normalized.split(/\s*,?\s+then\s+/);
  const steps = [];
  for (const clause of clauses) {
    let action = clause, gate;
    const before = clause.match(/^when (.+?),\s*(.+)$/);
    const after = clause.match(/^(.+?) when (.+)$/);
    if (before) {gate = condition(before[1]); action = before[2];}
    else if (after) {action = after[1]; gate = condition(after[2]);}
    if (gate) steps.push({op:'condition', value:gate, timeoutMs:60000});
    steps.push(command(action));
  }
  if (steps.length > MAX_NODES) fail('order exceeds 24 steps including conditions');
  return validateGraph({version:1, id, start:'n1', maxTransitions:steps.length, nodes:steps.map((step, i) => ({id:`n${i+1}`, ...step, success:i === steps.length - 1 ? 'done' : `n${i+2}`, failure:'failed'}))});
}

File: Source/mod/Scripts/companion_orders.lua
-- Portable bounded graph executor. No UE calls, reflection, text evaluation, or timers.
-- Caller drives tick(monotonic milliseconds), supplies capabilities, and owns cleanup.
local M = {}
local terminals = {done=true, failed=true}
local values = {
    follow={player=true}, hold={position=true}, attack={current_target=true},
    role={frontline=true,rearguard=true,flank_left=true,flank_right=true,ranged=true},
    stance={passive=true,defensive=true,aggressive=true},
    priority={player_target=true,protect=true,nearest=true,weakest=true},
}
local function integer(n,lo,hi) return type(n)=='number' and n==math.floor(n) and n>=lo and n<=hi end
local function identifier(s) return type(s)=='string' and #s<=48 and s:match('^[A-Za-z][A-Za-z0-9_-]*$')~=nil end
local function validValue(op,value,timeout)
    if type(value)~='string' then return false end
    if values[op] then return values[op][value]==true end
    if op=='wait' then return value:match('^[1-9]%d*$')~=nil and #value<=6 and integer(tonumber(value),1,120000) and tonumber(value)<=timeout end
    if op=='condition' then
        if value=='combat' or value=='target_clear' then return true end
        local fraction=value:match('^health_below:(0%.%d+)$')
        return fraction~=nil and #fraction<=5 and tonumber(fraction)>0
    end
    return false
end
function M.validate(input)
    if type(input)~='table' or input.version~=1 or not identifier(input.id) or terminals[input.id] then return nil,'Invalid graph version or id' end
    if not integer(input.maxTransitions,1,96) then return nil,'Execution needs an explicit 1-96 transition bound' end
    if type(input.nodes)~='table' or #input.nodes<1 or #input.nodes>24 then return nil,'Provide 1-24 nodes' end
    local graph={version=1,id=input.id,start=input.start,maxTransitions=input.maxTransitions,nodes={},byId={}}
    local count=0
    for key in pairs(input.nodes) do
        if not integer(key,1,#input.nodes) then return nil,'Nodes must be a dense array' end
        count=count+1
    end
    if count~=#input.nodes then return nil,'Nodes must be a dense array' end
    for i,n in ipairs(input.nodes) do
        if type(n)~='table' or not identifier(n.id) or terminals[n.id] or graph.byId[n.id] then return nil,'Invalid or duplicate node id' end
        if not integer(n.timeoutMs,1,120000) or not validValue(n.op,n.value,n.timeoutMs) then return nil,'Unsupported operation/value or timeout at '..n.id end
        if type(n.success)~='string' or type(n.failure)~='string' then return nil,'Both edges are required at '..n.id end
        local copy={id=n.id,op=n.op,value=n.value,timeoutMs=n.timeoutMs,success=n.success,failure=n.failure}
        graph.nodes[i]=copy;graph.byId[n.id]=copy
    end
    if not graph.byId[graph.start] then return nil,'Missing start node' end
    for _,n in ipairs(graph.nodes) do
        if not terminals[n.success] and not graph.byId[n.success] then return nil,'Missing success edge at '..n.id end
        if not terminals[n.failure] and not graph.byId[n.failure] then return nil,'Missing failure edge at '..n.id end
    end
    local reachable={}
    local function visit(id)
        if terminals[id] or reachable[id] then return end
        reachable[id]=true;local n=graph.byId[id];visit(n.success);visit(n.failure)
    end
    visit(graph.start)
    local exits={done=true,failed=true}
    for _=1,#graph.nodes do for _,n in ipairs(graph.nodes) do if exits[n.success] or exits[n.failure] then exits[n.id]=true end end end
    for _,n in ipairs(graph.nodes) do
        if not reachable[n.id] then return nil,'Unreachable node '..n.id end
        if not exits[n.id] then return nil,'Closed loop at '..n.id end
    end
    return graph
end
local function split(line)
    local parts={};for value in (line..'\t'):gmatch('(.-)\t') do parts[#parts+1]=value end;return parts
end
function M.decode(text)
    if type(text)~='string' or #text>12288 or text:find('[^\9\10\13\32-\126]') then return nil,'Invalid or oversized wire data' end
    text=text:gsub('\r\n','\n'):gsub('\n$','')
    local lines={};for line in (text..'\n'):gmatch('(.-)\n') do lines[#lines+1]=line end
    local h=split(lines[1] or '')
    if #h~=5 or h[1]~='ORDER' or h[2]~='1' or not h[5]:match('^[1-9]%d?$') then return nil,'Invalid wire header' end
    local graph={version=1,id=h[3],start=h[4],maxTransitions=tonumber(h[5]),nodes={}}
    if #lines>25 then return nil,'Too many wire nodes' end
    for i=2,#lines do
        local p=split(lines[i])
        if #p~=7 or p[1]~='NODE' or not p[5]:match('^[1-9]%d*$') or #p[5]>6 then return nil,'Invalid wire node' end
        graph.nodes[#graph.nodes+1]={id=p[2],op=p[3],value=p[4],timeoutMs=tonumber(p[5]),success=p[6],failure=p[7]}
    end
    return M.validate(graph)
end

local Runner={};Runner.__index=Runner
function Runner:snapshot()
    return {id=self.graph.id,status=self.status,nodeId=self.nodeId,transitions=self.transitions,message=self.message,active=self.active~=nil}
end
function Runner:release(reason)
    if self.active and self.dispatched and type(self.adapter.cancel)=='function' then pcall(self.adapter.cancel,self.active.id,reason) end
    self.active=nil;self.dispatched=false
end
function Runner:cancel(reason)
    if self.status=='running' then self:release(reason or 'Cancelled');self.status='cancelled';self.message=reason or 'Cancelled' end
    return self:snapshot()
end
function Runner:transition(success,message)
    local edge=success and self.active.success or self.active.failure
    if not success then self:release(message or 'Action failed') else self.active=nil;self.dispatched=false end
    self.message=message or (success and 'Step completed' or 'Step failed')
    self.nodeId=edge
    if terminals[edge] then self.status=edge end
end
local function result(value)
    if value==true or value=='success' then return 'success' end
    if value==false or value=='failure' then return 'failure' end
    if value=='pending' then return 'pending' end
    return 'unsupported'
end
function Runner:tick(now)
    if self.status~='running' then return self:snapshot() end
    if type(now)~='number' or now~=now or now==math.huge or now<self.lastNow then
        self:release('Invalid clock');self.status='failed';self.message='Clock must be finite and monotonic';return self:snapshot()
    end
    self.lastNow=now
    if not self.active then
        if self.transitions>=self.graph.maxTransitions then self.status='failed';self.message='Transition budget exhausted';return self:snapshot() end
        self.transitions=self.transitions+1;self.active=self.graph.byId[self.nodeId];self.started=now;self.dispatched=false
    end
    local n=self.active;local elapsed=now-self.started
    if n.op=='wait' then
        if elapsed>=tonumber(n.value) then self:transition(true,'Wait completed') end
    elseif n.op=='condition' then
        local name,value=n.value:match('^([^:]+):(.+)$');name=name or n.value;value=value and tonumber(value) or nil
        if type(self.adapter.sense)~='function' then self:transition(false,'Condition sensing unavailable')
        else
            local ok,ready,why=pcall(self.adapter.sense,name,value)
            if not ok then self:transition(false,'Condition adapter failed')
            elseif ready==true then self:transition(true,'Condition met')
            elseif ready~=false then self:transition(false,why or 'Condition unsupported')
            elseif elapsed>=n.timeoutMs then self:transition(false,'Condition timed out') end
        end
    else
        local ok,value,why
        if not self.dispatched then
            self.dispatched=true
            if type(self.adapter.action)=='function' then ok,value,why=pcall(self.adapter.action,n.op,n.value,n.id)
            else ok=true;value=nil;why='Action adapter unavailable' end
        elseif elapsed>=n.timeoutMs then
            self:transition(false,'Action timed out');return self:snapshot()
        elseif type(self.adapter.poll)=='function' then ok,value,why=pcall(self.adapter.poll,n.id,n.op,n.value)
        else return self:snapshot() end
        local status=ok and result(value) or 'error'
        if status=='success' then self:transition(true,why or 'Action accepted')
        elseif status~='pending' then self:transition(false,why or (status=='error' and 'Action adapter failed' or 'Action unsupported or failed')) end
    end
    -- At most one node is entered per tick. A caller can tick slowly without busy loops.
    return self:snapshot()
end
function M.new(input,adapter,now)
    local graph,err=M.validate(input);if not graph then return nil,err end
    if type(adapter)~='table' then return nil,'Adapter is required' end
    now=now or 0
    if type(now)~='number' or now~=now or now==math.huge or now==-math.huge then return nil,'Clock must be finite' end
    return setmetatable({graph=graph,adapter=adapter,status='running',nodeId=graph.start,transitions=0,lastNow=now,message='Ready',dispatched=false},Runner)
end
return M
